overlord/restart/restart_test.go

```diff
@@ -660,6 +686,53 @@ func (s *restartSuite) TestTaskWaitForRestartInvalid(c *C) {
 	c.Assert(err, ErrorMatches, `internal error: only tasks currently in progress \(doing/undoing\) are supported`)
 }
 
+func (s *restartSuite) TestFinishTaskWithRestart2DeferredRestart(c *C) {
+	restore := release.MockOnClassic(false)
+	defer restore()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	_, err := restart.Manager(st, "boot-id-1", nil)
+	c.Assert(err, IsNil)
+
+	chg1 := st.NewChange("test", "...")
+	t1 := st.NewTask("restarting-task", "...")
+	t2 := st.NewTask("descending-task", "...")
+	chg1.AddTask(t1)
+	chg1.AddTask(t2)
```

Comment:
Do we need the descending task in this test?
```diff
diff --git a/overlord/restart/restart_test.go b/overlord/restart/restart_test.go
index 9706f06506..665916af30 100644
--- a/overlord/restart/restart_test.go
+++ b/overlord/restart/restart_test.go
@@ -699,17 +699,13 @@ func (s *restartSuite) TestFinishTaskWithRestart2DeferredRestart(c *C) {
 
        chg1 := st.NewChange("test", "...")
        t1 := st.NewTask("restarting-task", "...")
-       t2 := st.NewTask("descending-task", "...")
        chg1.AddTask(t1)
-       chg1.AddTask(t2)
-       t2.WaitFor(t1)
 
        restart.SetDeferredRestartForTask(t1)
 
        err = restart.FinishTaskWithRestart2(t1, "some-snap", state.DoneStatus, restart.RestartSystemNow, nil)
        c.Assert(err, IsNil)
        c.Check(t1.Status(), Equals, state.DoneStatus)
-       c.Check(t2.Status(), Equals, state.DoStatus)
 
        rt, err := restart.RestartParametersFromChange(chg1)
        c.Assert(err, IsNil)
```

Reply:
Really good question, however I choose to keep one task in each direction in, to ensure that their status'es are unaffected by the `FinishTaskWithRestart2`. So it's for completeness and my opinion is that they add value. Especially since without the `SetDeferredRestartForTask` call the statuses for their descending tasks (in each direction) would change.

Reply:
Makes sense, Thank you for explaining!